Documented RBAC and Updated Authorization Documentation

greg-szrama opened the pull request:
Resolves: https://github.com/redpanda-data/documentation-private/issues/2235

Created a new RBAC guide and placed it under a new Authorization section under Manage -> Security.

Preview links.
[Redpanda Authorization.](https://deploy-preview-373--redpanda-docs-preview.netlify.app/24.1/manage/security/authorization/)
[Access Control Lists](https://deploy-preview-373--redpanda-docs-preview.netlify.app/24.1/manage/security/authorization/acl/)
[Role-based Access Control](https://deploy-preview-373--redpanda-docs-preview.netlify.app/24.1/manage/security/authorization/rbac/)

mattschumpert commented:
@deniscoady should review @greg-szrama

mattschumpert commented:
Please put the preview link in the description @greg-szrama

greg-szrama commented:
> Please put the preview link in the description @greg-szrama

My apologies, @mattschumpert , that's fixed!

greg-szrama commented:
Note for reviewers - I'm setting up a separate PR to include the Console configurations and I'll remove the Console information from this document.

oleiman commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,131 @@
+
+[,bash]
+----
+rpk acl create ... --allow-principal Role:[ROLE_NAME]
```
This is a bit stale 

```suggestion
rpk acl create ... --allow-role [ROLE_NAME]
```
Please lmk if there are docs hanging around that still contain this usage. Keeping them all up to date is...challenging 😅

oleiman commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,131 @@
+
+=== Edit a Role
+
+After a role is created, administrators may still add additional ACLs to the role. Redpanda also allows renaming a role.
+
+Using RPK, Administrators may rename a role by issuing the command:
+
+[,bash]
+----
+rpk role edit [ROLE_NAME] --rename [NEW_NAME]
+----
```
This functionality (role renaming) has been abandoned.

greg-szrama replied:
Removed the reference.

oleiman commented on modules/manage/pages/security/authorization/acl.adoc:
```diff
@@ -1,14 +1,14 @@
 
 
-Authorization works in tandem with xref:./authentication.adoc[authentication]. Access-control lists (ACLs) provide a way to configure fine-grained access to provisioned users. ACLs work with SASL/SCRAM and with mTLS with principal mapping for authentication.
+Access-control lists (ACLs) provide a way to configure fine-grained access to provisioned users. ACLs work with SASL/SCRAM and with mTLS with principal mapping for authentication.
```
unrelated to this PR, but I've noticed it before. ACLs also work with OIDC and GSSAPI principal mapping. Might be worth changing in this round. @deniscoady?

greg-szrama replied:
@deniscoady - any input on this? Do we need to change anything on the language here?

oleiman commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,131 @@
+= Configure Role-based Access Control
+:description: Role-based access controls provide an alternative to ACLs for managing permissions.
```
nitpick: This feels tricky to express, but RBAC is not technically an _alternative_ to ACLs. You still need the ACLs, but instead of binding them to individual users, you bind ACLs to a role and assign users to the role. Idk if that's a meaningful distinction for public docs though.

greg-szrama replied:
What if I rephrase it as, "...provides an extension to ACLs for managing permissions at scale."

oleiman replied:
LGTM

oleiman commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,131 @@
+
+RBAC is designed to address the critical challenge of access management at scale. It alleviates the process of manually maintaining and verifying a set of raw permissions (ACLs) for a user base that may contain thousands of users. RBAC provides you with a method to onboard new employees easier, audit accesses faster, and adapt more readily to evolutions of usage and compliance needs.
+
+Using RBAC, you can define roles to reflect organizational structure or job duties to gives a coherent approach to access management. This approach decouples users and permissions through the assignment of roles. This allows a one to many mapping of a given role to many users, thereby dramatically reducing the number of custom policies needed for a resource from one per user to one per group of u ...
```
>to gives a coherent approach

phrasing

oleiman commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,131 @@
+A role is a named collection of ACLs which can be assigned to zero or more users. Redpanda provides two default roles, `user` and `admin`:
+
+* `user` - All authorized users have this role by default. There are no ACLs assigned to this role until you add them. This means users have no permissions to interact with any resources you add to your Redpanda cluster until you add relevant permissions to this or another role.
+* `admin` - By default, no users have this role. The default admin role provides full access to all resources in your Redpanda installation.
```
This is not planned afaik. We may add it in the future, but not for beta or MVP. @deniscoady - can you confirm?

oleiman commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,131 @@
+
+A role is a named collection of ACLs which can be assigned to zero or more users. Redpanda provides two default roles, `user` and `admin`:
+
+* `user` - All authorized users have this role by default. There are no ACLs assigned to this role until you add them. This means users have no permissions to interact with any resources you add to your Redpanda cluster until you add relevant permissions to this or another role.
```
`s/user/Users/` per PRD, though I'm not opinionated on naming.

oleiman replied:
>authorized users have this role

Authorization status shouldn't come into play here. As I understand it, this should just apply to SASL/SCRAM users existing in the system at the time of cluster upgrade to an RBAC-supporting version of redpanda.

@deniscoady  - is this supposed to apply to every SASL user created, in perpetuity?

greg-szrama replied:
I can update it to 'Users' for the role name. Is the other 'Admins' then? I was unclear as to whether that was the explicit role name or a plural reference to that type of account. Can we confirm that it was implemented pluralized as the PRD specified it @oleiman ?

oleiman replied:
Confirmed it is "Users" in the PRD. "Admin" has been dropped and will not be included.

Also confirmed with Denis that we will assign this role ONLY to already extant SASL/SCRAM users _at the time of upgrade_. I still suspect that some finesse from product is required with respect to messaging.

Feediver1 replied:
The doc should match whatever is in the code/PRD. Pick one and be consistent. This should be decided by @deniscoady.

oleiman commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,131 @@
+
+== Managing users and roles
+
+Redpanda allows you to manage your RBAC configurations through both RPK and the web console. Examples of each are given here.
```
>both RPK and the web console

Conflicts with NOTE in index, which indicates "rpk only for now". @deniscoady - what is the preferred messaging here?

greg-szrama replied:
This was an oversight from when I was stripping out the content giving steps for Console interactions. I'll fix it!

oleiman commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,131 @@
+
+=== Create a role
+
+Creating a new role is a two-step process. First you define the role, giving it a unique and descriptive name. Second, assign that role as the allowed or denied principal to one or more ACLs. This defines the permissions associated with the role.
```
>allowed or denied principal

I think we're trying to avoid referring to the role as a "principal" as much as possible. For example, we don't want people to expect nested roles (i.e. a role assigned to a role)

greg-szrama replied:
I rephrased this so that it doesn't refer to principals.

oleiman reviewed (commented):
looking good! most of my comments are actually questions for product.

Feediver1 commented on modules/manage/pages/security/authorization/index.adoc:
```diff
@@ -0,0 +1,13 @@
+
+[NOTE]
+====
+At the present time, RBAC is only available for Redpanda self-hosted users. You can only configure RBAC through the use of command-line RPK commands.
```
```suggestion
RBAC is only available for Redpanda self-hosted users. You can only configure RBAC through the use of command-line RPK commands.
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+:description: Role-based access controls provide an extension to ACLs for managing permissions at scale.
+:page-categories: Management, Security
+
+Role-based Access Control (RBAC) provides a way to configure permissions for provisioned users at scale. RBAC allows you to manage access for many users to many resources with a streamlined interface. This mechanism works in conjunction with all supported authentication methods.
```
```suggestion
Role-based Access Control (RBAC) provides a way to configure permissions for provisioned users at scale, and provides a streamlined interface to manage access for many users to many resources. RBAC works in conjunction with all supported authentication methods.
```

oleiman replied:
🔥

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+Role-based Access Control (RBAC) provides a way to configure permissions for provisioned users at scale. RBAC allows you to manage access for many users to many resources with a streamlined interface. This mechanism works in conjunction with all supported authentication methods.
+
+IMPORTANT: At present, Redpanda only supports RBAC for self-hosted clusters. Configuration is only supported through the command line using RPK.
```
```suggestion
IMPORTANT: Redpanda only supports RBAC for self-hosted clusters. Configuration is only supported through the command line using RPK.
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+== RBAC
+
+RBAC is designed to address the critical challenge of access management at scale. It alleviates the process of manually maintaining and verifying a set of raw permissions (ACLs) for a user base that may contain thousands of users. RBAC provides you with a method to onboard new employees easier, audit accesses faster, and adapt more readily to evolutions of usage and compliance needs.
```
```suggestion
RBAC is designed to address the critical challenge of access management at scale. It alleviates the process of manually maintaining and verifying a set of raw permissions (ACLs) for a user base that may contain thousands of users. RBAC provides a method to onboard new employees easier, audit accesses faster, and adapt more readily to evolutions of usage and compliance needs.
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+RBAC is designed to address the critical challenge of access management at scale. It alleviates the process of manually maintaining and verifying a set of raw permissions (ACLs) for a user base that may contain thousands of users. RBAC provides you with a method to onboard new employees easier, audit accesses faster, and adapt more readily to evolutions of usage and compliance needs.
+
+Using RBAC, you can define roles to reflect organizational structure or job duties to implement a coherent approach to access management. This approach decouples users and permissions through the assignment of roles. This allows a one to many mapping of a given role to many users, thereby dramatically reducing the number of custom policies needed for a resource from one per user to one per group  ...
```
```suggestion
Using RBAC, you can define roles to reflect organizational structure or job duties while implementing a coherent approach to access management. This approach decouples users and permissions through the assignment of roles. RBAC allows a one-to-many mapping of a given role to many users, dramatically reducing the number of custom policies needed for a resource from one per user to one per group of users.
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+Using RBAC, you can define roles to reflect organizational structure or job duties to implement a coherent approach to access management. This approach decouples users and permissions through the assignment of roles. This allows a one to many mapping of a given role to many users, thereby dramatically reducing the number of custom policies needed for a resource from one per user to one per group  ...
+
+image::shared:rbac-overview.png[RBAC multiplicity]
```
The image kind of just appears--you should refer to it in the narrative. 
eg. The following diagram shows user-to-role mappings using RBAC.
Also--the resolution seems off--the text on the right looks blurry to me. Advise against using internal names like this. Perhaps User1, User 2, and so on. Check with @deniscoady

Feediver1 replied:
May want to add principal to our glossary too.

greg-szrama replied:
We're going to need more input from @deniscoady  and @oleiman on this front. I was trying to make the RBAC narrative consistent with the writeups in ACLs which talk about Principals, not Users. In Redpanda, it seems the only type of Principal we have is Users anyway, at least based on what I've seen in the ACL documentation, the PRD, and elsewhere. I defer to Denis on this since we also heard that they want to minimize the use of the term Principal to avoid confusion.

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+=== RBAC terminology
+
+Under the RBAC framework, you create *roles*, grant *permissions* to those roles, and assign the roles to *users*. When you change the permissions for a given role, all principals with that role automatically gain the modified permissions. You grant or deny permissions for a role by creating an ACL and specifying the RBAC role as the allowed or denied principal respectively.
```
You grant or deny permissions for a role by creating an ACL and specifying the RBAC role as the allowed or denied principal respectively.
^^ Super important point--how about a specific example here? For example,...

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+==== Roles
+
+A role is a named collection of ACLs which can be assigned to zero or more users. Redpanda provides two default roles, `user` and `admin`:
```
```suggestion
A role is a named collection of ACLs that can be assigned to zero or more users. Redpanda provides two default roles, `user` and `admin`:
```

Feediver1 replied:
How/when would one assign a role to zero users?

oleiman replied:
You wouldn't, but there might exist a role which is not currently assigned to any user principals. An "empty" role is perfectly legal.

greg-szrama replied:
I'll rephrase -- instead of "can be assigned to zero or more users", "that has zero or more assigned users". I'll also rephrase the bit about two default roles in light of the removal of 'Admins' and the fact that 'Users' is only there for upgraded clusters, not fresh clusters.

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+A role is a named collection of ACLs which can be assigned to zero or more users. Redpanda provides two default roles, `user` and `admin`:
+
+* `user` - All authorized users have this role by default. There are no ACLs assigned to this role until you add them. This means users have no permissions to interact with any resources you add to your Redpanda cluster until you add relevant permissions to this or another role.
+* `admin` - By default, no users have this role. The default admin role provides full access to all resources in your Redpanda installation.
```
```suggestion
* `admin` - By default, no users are assigned this role. The default admin role provides full access to all resources in your Redpanda installation.
```

Feediver1 replied:
@oleiman So the admin role is for machines/apps to have access? Should we explain more explicitly than saying "all resources"?

Feediver1 replied:
Ah--just saw your comment saying it is not going in the release...how would apps use/access RP without RBAC roles? ACLs only?

oleiman replied:
Superuser access is available, as ever, for comprehensive "do anything" authorization. Superusers can then create roles, bind ACLs to those roles, assign user principals, etc.

greg-szrama replied:
When I remove the 'Admin' role description, I'll add a note here that superuser access is required for establishing roles until users establish their own administrator roles.

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+==== Policy conflicts
+
+You can assign a combination of ACLs and roles to any given principal. ACLs can contain allow permissions, deny permissions, or a combination of both. As a result, users may at times have conflicting policies. If this situation arises, the user is permitted to perform an operation if and only if:
```
```suggestion
You can assign a combination of ACLs and roles to any given principal. ACLs allow permissions, deny permissions, or specify a combination of both. As a result, users may at times have role assignments with conflicting policies. If this situation arises, the user is permitted to perform an operation if and only if:
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+== Managing users and roles
+
+Redpanda allows you to manage your RBAC configurations through both RPK. Examples of each are given here.
```
```suggestion
Redpanda allows you to manage your RBAC configurations using rpk. 
```

Feediver1 replied:
Not sure what the other method you intended to introduce here--Admin API? You never included it so I revised. I know we have Admin API updates for RBAC. If that is what you intended here, please add and sync with @kbatuigas on links.

greg-szrama replied:
Oh shoot. Originally this had a writeup for RPK or Console, but Console was removed from the beta. I did a quick scrub of alternatives but missed this. There will be Admin API updates, and I've already spoken briefly with @kbatuigas on it, I intend to have a deeper discussion next week with her. As far as I'm aware, Admin API is not ready for beta either, but we can confirm this with @deniscoady and @oleiman .

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+Creating a new role is a two-step process. First you define the role, giving it a unique and descriptive name. Second, you assign one or more ACLs to allow or deny access for the new role. This defines the permissions that are inherited by all users assigned to the role.
+
+Using RPK, administrators create a new role by issuing the command:
```
```suggestion
Administrators can use rpk to create a new role by issuing the command:
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+rpk role create [ROLE_NAME]
+----
+
+Once the role is created, administrators assign the role as the allowed principal for one or more ACLs. You can modify existing ACLs or create new ones. For example:
```
```suggestion
After the role is created, administrators assign the role as the allowed principal for one or more ACLs. You can modify existing ACLs or create new ones. For example:
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+When a role is deleted, Redpanda carries out the following actions automatically:
+
+- All role ACLs are deleted.
+- All users are unassigned from role.
```
```suggestion
- All users are unassigned the role.
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+- All role ACLs are deleted.
+- All users are unassigned from role.
+
+Using RPK, administrations delete a role by issuing the command:
```
```suggestion
Using rpk, administrations delete a role by issuing the command:
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+=== Assign a role
+
+Administrators may assign a role to any principal created in their Redpanda instance. A principal is composed of two parts: a type and a name. Redpanda currently supports the `User` type only for principals. All principals, therefore, will have the format of `User:[name]`
```
```suggestion
Administrators may assign a role to any principal created in their Redpanda instance. A principal is composed of two parts: a type and a name. Redpanda supports the `User` type only for principals. All principals, therefore, have the format: `User:[name]`
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+Administrators may assign a role to any principal created in their Redpanda instance. A principal is composed of two parts: a type and a name. Redpanda currently supports the `User` type only for principals. All principals, therefore, will have the format of `User:[name]`
+
+Using RPK, administrations assign a role to a principal by issuing the command:
```
```suggestion
Using rpk, administrations assign a role to a principal by issuing the command:
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+=== Unassign a role
+
+Administrators may remove a role from a principal without deleting the role. A principal is composed of two parts: a type and a name. Redpanda currently supports the `User` type only for principals. All principals, therefore, will have the format of `User:[name]`
```
Administrators may remove a role from a principal without deleting the role. A principal is composed of two parts: a type and a name. Redpanda currently supports the `User` type only for principals. All principals, therefore, will have the format of `User:[name]`
^^ This should come earlier in the doc--please see my earlier comment asking how users and principals differ here...Perhaps put **principal** in the terminology section?

greg-szrama replied:
@oleiman  and @deniscoady  -- I was using this in part from the ACL documentation which talks more about Principals and Users. @oleiman mentioned that we're shying away from talking about Principals. Do you want me to remove any reference to Principals in the ACL doc and just say, you specify the user with the format `User:[name]`?

oleiman replied:
Mentioned in slack but, for completeness, the important distinction is "(User) Principals" vs "Roles".

The "principal" terminology is not problematic when describing ACLs (unless the ACL applies to a role)

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+Administrators may remove a role from a principal without deleting the role. A principal is composed of two parts: a type and a name. Redpanda currently supports the `User` type only for principals. All principals, therefore, will have the format of `User:[name]`
+
+Using RPK, administrators unassign a role from a principal issuing the command:
```
```suggestion
Using rpk, administrators unassign a role from a principal issuing the command:
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+=== Unassign a role
+
+Administrators may remove a role from a principal without deleting the role. A principal is composed of two parts: a type and a name. Redpanda currently supports the `User` type only for principals. All principals, therefore, will have the format of `User:[name]`
```
```suggestion
Administrators may remove a role from a principal without deleting the role. A principal is composed of two parts: a type and a name. Redpanda currently supports the `User` type only for principals. All principals, therefore, have the format: `User:[name]`
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+=== Edit a Role
+
+After a role is created, administrators may still add additional ACLs to the role.
```
```suggestion
Administrators can modify an existing role by adding additional ACLs to it.
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+After a role is created, administrators may still add additional ACLs to the role.
+
+Using RPK, Administrators may assign the role as the allowed principal for additional ACLs. You can modify existing ACLs or create new ones. For example:
```
```suggestion
Using rpk, administrators may assign the role as the allowed principal for additional ACLs. You can modify existing ACLs or create new ones. For example:
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+Administrators may view a list of all roles within their Redpanda installation.
+
+Using RPK, administrators can view a list of all actives roles by issuing the command:
```
```suggestion
Using rpk, administrators can view a list of all actives roles by issuing the command:
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,124 @@
+
+Administrators may view all principals with a given role assignment.
+
+Using RPK, administrators can view the membership of a given role by issuing the command:
```
```suggestion
Using rpk, administrators can view the membership of a given role by issuing the command:
```

greg-szrama commented:
@oleiman one question on the 'delete' role capability -- the PRD said it deletes all role-specific ACLs. Does this delete all ACLs the role is assigned to, or only the ACLs that have *only* this role assigned to them?

oleiman commented:
@greg-szrama - ACLs are always 1:1 with a role or user, so it's always the case that an ACL assigned to a role is only assigned to that role. Put another way, the two alternatives in your question are equivalent, so the answer is "both" 🙂

greg-szrama commented:
I've resolved initial feedback and this is ready for full review. I've addressed some issues with other documents, especially the ACLs document. I added the `--allow-role` and `--deny-role` options to the ACL documentation. Please review @oleiman , @deniscoady , and @Feediver1 . @kbatuigas , we'll need to work in the Admin API references once we go live with your work.

r-vasquez commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+----
+rpk security role create 'Software Eng.'
+rpk acl create --operation read --operation write --topic private-* --allow-role 'Software Eng.'
+rpk security role assign 'Software Eng.' --principal User:mike
+rpk security role assign 'Software Eng.' --principal User:oren
```
The `--principal` flag accepts multiple values, so you can do:

```
rpk security role assign 'Software Eng.' --principal User:mike,User:oren
``` 

Also, we can document this without the `User:`, rpk will assume the principal type `User:` if no type is provided:
```
rpk security role assign 'Software Eng.' --principal mike,oren
```

r-vasquez commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+
+image::shared:rbac-overview.png[RBAC multiplicity]
+
+Now consider the situation where you want to create a new topic, `private-software-versions`, where users self-report the version of a component they are using. When using the ACL authorization mechanism, you would need to assign this ACL to every user in your Redpanda installation. Using RBAC allows you to make one update and apply it to everyone with that role. Adding the write permissions for  ...
+
+[,bash]
+----
+rpk acl create --operation write --topic private-software-versions --allow-role User
+----
```
I'm confused with:

>Adding the write permissions for this topic to the `User` role means everyone with that role (which is all users in our example) gains the access immediately.

In the example above, the role was named `'Software Eng.'`, but here you mention the role `User`.

greg-szrama replied:
Yep! I'll update this to refer to the image and say something along the lines of, "as an extension of our example."

greg-szrama replied:
@r-vasquez please review the language here now. I've updated it for clarity.

r-vasquez commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+
+[,bash]
+----
+rpk security role assign [ROLE_NAME] [PRINCIPAL]
```
The flag is missing here:
```suggestion
rpk security role assign [ROLE_NAME] --principal [PRINCIPALS...]
```

r-vasquez commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+
+[,bash]
+----
+rpk security role unassign [ROLE_NAME] [PRINCIPAL]
```
```suggestion
rpk security role unassign [ROLE_NAME] --principal [PRINCIPALS...]
```

r-vasquez commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+rpk security role list
+----
+
+=== View Role Membership
+
+Using rpk, administrators can view the membership of a given role by issuing the command:
+
+[,bash]
+----
+rpk security role users [ROLE_NAME]
+----
```
This command does not exist in rpk, instead, we have `rpk security role describe [ROLE_NAME]`

```
$ rpk security role describe red
PERMISSIONS
===========
PRINCIPAL         HOST  RESOURCE-TYPE  RESOURCE-NAME  RESOURCE-PATTERN-TYPE  OPERATION  PERMISSION  ERROR
RedpandaRole:red  *     TOPIC          asdf           LITERAL                ALL        ALLOW
RedpandaRole:red  *     TOPIC          foo            LITERAL                ALL        ALLOW

PRINCIPALS (1)
==============
NAME  TYPE
foo   User
```

The PR contains examples of every command that was added: https://github.com/redpanda-data/redpanda/pull/17538

kbatuigas commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+rpk acl create --operation write --topic private-software-versions --allow-role User
+----
+
+== Managing users and roles
```
```suggestion
== Manage users and roles
```

kbatuigas commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+
+== Managing users and roles
+
+Redpanda allows you to manage your RBAC configurations through the use of RPK. The Redpanda admin API provides another means of interacting with RBAC. Examples of using rpk for each RBAC operation are given here.
```
```suggestion
Redpanda allows you to manage your RBAC configurations through the use of `rpk`. The Redpanda Admin API provides another means of interacting with RBAC. Examples of using `rpk` for each RBAC operation are given here.
```

kbatuigas commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+
+Administrators may remove a role assignment from a principal without deleting the role. Principals are referred to using the format: `Type:Name`. Redpanda currently supports only the `User` type. If you omit the type, Redpanda assumes the `User` type by default.
+
+Using rpk, administrators remove a role assignment from a principal by issuing the command:
```
```suggestion
Using `rpk`, administrators remove a role assignment from a principal by issuing the command:
```

kbatuigas commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+rpk security role unassign [ROLE_NAME] [PRINCIPAL]
+----
+
+=== Edit a Role
```
```suggestion
=== Edit a role
```

kbatuigas commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+
+=== Edit a Role
+
+Using rpk, administrators can modify an existing role by adding additional ACLs to it using the command:
```
```suggestion
Using `rpk`, administrators can modify an existing role by adding more ACLs to it using the command:
```
Suggestion if you'd like to avoid using very similar words right next to each other :)

kbatuigas commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+rpk acl create ... --allow-role [ROLE_NAME]
+----
+
+See xref:security/authorization/acl.adoc[Access Control Lists] for more information on defining ACLs.
```
```suggestion
See xref:security/authorization/acl.adoc[access control lists] for more information on defining ACLs.
```
Looks like we have standardized usage based on the style guide? https://github.com/redpanda-data/docs-site/blob/main/meta-docs/STYLE-GUIDE.adoc#word-list

kbatuigas commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+
+See xref:security/authorization/acl.adoc[Access Control Lists] for more information on defining ACLs.
+
+=== List Roles
```
```suggestion
=== List roles
```

kbatuigas commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+
+=== List Roles
+
+Using rpk, administrators can view a list of all actives roles by issuing the command:
```
```suggestion
Using `rpk`, administrators can view a list of all actives roles by issuing the command:
```

kbatuigas commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+rpk security role list
+----
+
+=== View Role Membership
```
```suggestion
=== View role membership
```

kbatuigas commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+
+=== View Role Membership
+
+Using rpk, administrators can view the membership of a given role by issuing the command:
```
```suggestion
Using `rpk`, administrators can view the membership of a given role by issuing the command:
```

kbatuigas reviewed (changes requested):
Just some minor suggested edits per the style guide

Deflaimun commented:
Updated preview links to 24.1

micheleRP commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+= Configure Role-based Access Control
```
```suggestion
= Configure Role-Based Access Control
```
@kbatuigas checking that you're OK with this!

greg-szrama replied:
@Feediver1 thoughts on capitalization here?

greg-szrama commented:
@oleiman @deniscoady @kbatuigas I've made some substantial updates based on feedback. I added in the samples for each RBAC command, as well as some better documentation on the `Edit role permissions` section, discussing the use of the `rpk acl create`, `rpk acl delete`, and `rpk acl list` commands.

r-vasquez commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,247 @@
+
+== Manage users and roles
+
+Redpanda allows you to manage your RBAC configurations through the use of `rpk`. The Redpanda Admin API provides another means of interacting with RBAC. Examples of using `rpk` for each RBAC operation are given here.
```
What are the other means of interacting with RBAC through the Admin API? Perhaps we can drop:

```
The Redpanda Admin API provides another means of interacting with RBAC
```

And reference the RBAC endpoints of the Admin API at the end of this page. wdyt?

greg-szrama replied:
Yep, that's fair -- I plan to work in references to admin API once @kbatuigas  gets her PR merged. We'll probably do that as a separate smaller PR after this one clears.

r-vasquez commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,247 @@
+PERMISSIONS
+===========
+PRINCIPAL         HOST  RESOURCE-TYPE  RESOURCE-NAME  RESOURCE-PATTERN-TYPE  OPERATION  PERMISSION  ERROR
+RedpandaRole:red  *     TOPIC          asdf           LITERAL                ALL        ALLOW
```
I remember @Feediver1 asking us to drop the usage of random strings as 'asdf' in other PRs, maybe rename this example to `books` or something else :sweat_smile:

greg-szrama replied:
I scrubbed all the examples, thanks for calling that out!

r-vasquez commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,247 @@
+rpk acl list --allow-role [ROLE_NAME] --deny-role [ROLE_name]
+----
+
+See also:
+
+* xref:security/authorization/acl.adoc[Access Control Lists] for more information on defining and using ACLs.
+* xref:reference:rpk/rpk-acl/rpk-acl-create.adoc[]
+* xref:reference:rpk/rpk-acl/rpk-acl-delete.adoc[]
+* xref:reference:rpk/rpk-acl/rpk-acl-list.adoc[]
```
Should this section be at the end? or is there a reason why is above `List` and `Describe`?

greg-szrama replied:
These are references to the ACL modification commands within `rpk`, discrete from the role commands. It's part of editing/managing the ACLs on your Roles.

r-vasquez reviewed (commented):
The rpk examples look good to me, my only additional advice is to change the Role `Software Eng.` to `SoftwareEng` since some shells might need extra work to parse the white space and dots.

oleiman reviewed (commented):
looking good! most of my comments are technical terminology distinctions, some may not be worth addressing so feel free to 👍/👎 as appropriate.

kbatuigas reviewed (approved):
Didn't find anything blocking so I'll approve, but saw some good questions/suggestions from Oren.

kbatuigas commented:
@greg-szrama Admin API PR: https://github.com/redpanda-data/docs/pull/397 The xref should look something like `api:admin-api.html#tag--Roles`

oleiman commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,247 @@
+
+== RBAC
+
+RBAC is designed to address the critical challenge of access management at scale. It alleviates the process of manually maintaining and verifying a set of raw permissions (access control lists, or ACLs) for a user base that may contain thousands of users. RBAC provides a method to onboard new employees easier, audit accesses faster, and adapt more readily to evolutions of usage and compliance nee ...
+
+Using RBAC, you can define roles to reflect organizational structure or job duties, enabling implementation of a coherent approach to access management. This approach decouples users and permissions through the assignment of roles. RBAC allows a one-to-many mapping of a given role to many users, thereby dramatically reducing the number of custom policies needed for a resource from one per user to ...
```
praise: I keep coming back to this passage. great stuff

greg-szrama commented on modules/manage/pages/security/authorization/acl.adoc:
```diff
@@ -1,14 +1,14 @@
 == ACLs
 
-Access-control lists (ACLs) are the primary mechanism used by Redpanda to manage user permissions. Redpanda stores ACLs internally, replicated with https://raft.github.io/[Raft^] to provide the same consensus guarantees as your data. You can manage ACLs with xref:reference:rpk/rpk-acl.adoc[`rpk acl`].
+Access-control lists (ACLs) are one of the primary mechanisms Redpanda provides to manage user permissions. Redpanda stores ACLs internally, replicated with https://raft.github.io/[Raft^] to provide the same consensus guarantees as your data. You can manage ACLs with xref:reference:rpk/rpk-acl.adoc[`rpk acl`].
```
I'm torn here, because technically even RBAC falls back to ACLs to manage permissions, it just gives a container for grouping and assigning the ACLs. Are there other permissions management schemes in the works? Thoughts?

greg-szrama commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,247 @@
+
+image::shared:rbac-overview.png[RBAC multiplicity]
+
+Building off the diagram, consider the situation where you want to create a new topic, `private-software-versions`, where users self-report the version of a component they are using. If you were using the ACL authorization mechanism, you would need to assign this ACL to every user in your Redpanda installation. Using RBAC allows you to make one update and apply it to everyone with that role. Addi ...
```
I was only using the 'Users' role as a reference to the diagram, not as the auto-created Users role from the RBAC upgrade. If we need to change the role name in the diagram to avoid confusion, though, I'm open to that!

greg-szrama commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,247 @@
+= Configure Role-Based Access Control
+:description: Role-based access controls provide an extension to access control lists for managing permissions at scale.
```
Good catch!

greg-szrama commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,247 @@
+
+==== Roles
+
+A role is a named collection of ACLs which may have users assigned to it. You can assign any number of roles to a given user. When installing a new Redpanda cluster, no roles are provisioned by default. If you wish to use RBAC, you will need to create your first roles using your `superuser` account.
```
Where I settled was to use the 'security principal' term in the specific method descriptions, but to work up to using that term in the introductory narrative. I went ahead and added (security principals) to tie it together a bit better.

greg-szrama commented:
> The rpk examples look good to me, my only additional advice is to change the Role `Software Eng.` to `SoftwareEng` since some shells might need extra work to parse the white space and dots.

I have a request out to Ben Weinstein to update that image, renaming the role. As soon as I get that update, I'll swap it in and update the verbiage.

oleiman reviewed (approved):
Lgtm!

micheleRP commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,246 @@
+:description: Role-based access controls are an extension to access control lists for managing permissions at scale.
+:page-categories: Management, Security
+
+Role-based Access Control (RBAC) provides a way to configure permissions for provisioned users at scale, and provides a streamlined interface to manage user access to many resources. RBAC works in conjunction with all supported authentication methods.
```
```suggestion
Role-based access control (RBAC) provides a way to configure permissions for provisioned users at scale, and provides a streamlined interface to manage user access to many resources. RBAC works in conjunction with all supported authentication methods.
```

micheleRP commented on modules/ROOT/nav.adoc:
```diff
@@ -181,7 +181,9 @@
-*** xref:manage:security/authorization.adoc[Configure Authorization]
+*** xref:manage:security/authorization/index.adoc[Redpanda Authorization Mechanisms]
+**** xref:manage:security/authorization/acl.adoc[Access Control Lists]
+**** xref:manage:security/authorization/rbac.adoc[Role-based Access Control]
```
```suggestion
**** xref:manage:security/authorization/rbac.adoc[Role-Based Access Control]
```

micheleRP commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,246 @@
+
+This diagram shows the relationships between users, roles, and ACLs, including the example `Software Eng.` role:
+
+image::shared:rbac-overview.png[RBAC multiplicity]
```
This diagram is blurry.

micheleRP commented:
@greg-szrama Did you confirm if this page for [RBAC in Console (self-hosted)](https://docs.redpanda.com/current/manage/security/console/authorization/) needs updates?

greg-szrama commented:
> @greg-szrama Did you confirm if this page for [RBAC in Console (self-hosted)](https://docs.redpanda.com/current/manage/security/console/authorization/) needs updates?

From what I've gathered, this is specifically cloud console roles, for the management console itself. It's more an infrastructure role versus the roles to read/write to topics, the latter of which the current update implements. I do have some updates to this cloud console RBAC document staged right now. I planned to add those changes back in when we add in the steps to use cloud console to manage roles for users of the cluster. The idea was to clear up any potential confusion over where the two RBAC systems might seem to overlap.

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,146 @@
+
+RBAC is designed to address the critical challenge of access management at scale. It alleviates the process of manually maintaining and verifying a set of raw permissions (ACLs) for a user base that may contain thousands of users. RBAC provides a method to onboard new employees easier, audit accesses faster, and adapt more readily to evolutions of usage and compliance needs.
+
+Using RBAC, you can define roles to reflect organizational structure or job duties, enabling implementation of a coherent approach to access management. This approach decouples users and permissions through the assignment of roles. RBAC allows a one-to-many mapping of a given role to many users, thereby dramatically reducing the number of custom policies needed for a resource from one per user to ...
```
^^ is this suggestion being implemented?

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,282 @@
+When a role is deleted, Redpanda carries out the following actions automatically:
+
+- All role ACLs are deleted.
+- All users assignments to the role are removed.
```
```suggestion
- All users' assignments to the role are removed.
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,282 @@
+rpk security role delete <role-name>
+----
+
+Deleting a role, named `red`:
```
```suggestion
To delete the role named `red`:
```

Feediver1 commented on modules/manage/pages/security/authorization/rbac.adoc:
```diff
@@ -0,0 +1,282 @@
+rpk security role assign <role-name> --principal <principals>
+----
+
+Assigning a role, named "red":
```
```suggestion
Assign a role named "red":
```
